Set up docker image

cansavvy opened the pull request:
## Purpose:
Closes #158 

## Strategy
I borrowed items from the openPBTA Dockerfile, but trimmed out most of those packages.
We will need to add things as we go through each of the analysis modules (and add new analysis modules).

## Concerns/Questions for reviewers:
- I did not push this Docker image to Docker hub yet (want to get it approved first) so the docker pull step won't work. yet. 
- There are some plotting packages like `GGally` and `ggfortify` which we are not currently using, but I have in here, because I could see these being handy to have. 
- I am unsure which of these we still need, but I left them in for now: 
```
# Add curl, bzip2 and some dev libs
RUN apt-get update -qq && apt-get -y --no-install-recommends install \
    curl \
    bzip2 \
    zlib1g \
    libbz2-dev \
    liblzma-dev \
    libreadline-dev
```

## Pull Request Check List (roughly in order):
NA * [x] [Sources are cited](https://github.com/AlexsLemonade/refinebio-examples/CONTRIBUTING.md#citing-sources-in-text)
NA * [x] Set the seed (if applicable).
* [x] Remove any manual numbering of sections.
* [x] Remove any instances of chunk naming.
* [x] Spell check any Rmd file or md file.
* [x] Comments and documentation are up to date
NA  [x] Run [snakemake for rendering](https://github.com/AlexsLemonade/refinebio-examples/CONTRIBUTING.md#how-to-re-render-the-notebooks)

jashapiro commented on Dockerfile:
```diff
@@ -0,0 +1,75 @@
+    "cycler==0.10.0" "kiwisolver==1.1.0" "pyparsing==2.4.5" "python-dateutil==2.8.1" "pytz==2019.3" \
+    "numpy==1.17.3" \
+    "pandas==0.25.3" \
+    "snakemake==5.8.1" \
```
```suggestion
    "snakemake==5.8.1" 

# set final workdir for commands
WORKDIR /home/rstudio
```

jashapiro commented on CONTRIBUTING.md:
````diff
@@ -1,6 +1,38 @@
+
+Replace `<LOCAL_PATH_TO_REFINEBIO-EXAMPLES_REPO>` and run.
+```
+docker run -it --rm --mount type=volume,dst=/home/rstudio/kitematic,volume-driver=local,volume-opt=type=none,volume-opt=o=bind,volume-opt=device=<LOCAL_PATH_TO_REFINEBIO-EXAMPLES_REPO> -e PASSWORD=<PASSWORD> -p 8787:8787 refinebio-examples
````
I think we can/should simplify this. Notice that I am not using `/kitematic`, because I don't think it is needed, and just adds a level of confusion since you have to change directories to work with your local files. This way RStudio opens directly to the path you give. 
```suggestion
docker run --mount type=bind,target=/home/rstudio,source=<LOCAL_PATH_TO_REFINEBIO-EXAMPLES_REPO> -e PASSWORD=<PASSWORD> -p 8787:8787 refinebio-examples
```
We also should not need `-it` unless we are running a separate command and need to interact with the terminal, which we are not for general use. We also don't generally want to remove the image after we quit, so no `--rm`.

For the `snakemake` instructions, you could potentially add the following which will run `snakemake` in the correct directory without needing a web session. (`-it` could be used here, but should not be necessary)
```
docker run -w /home/rstudio --mount type=bind,target=/home/rstudio,source=<LOCAL_PATH_TO_REFINEBIO-EXAMPLES_REPO> -e  refinebio-examples snakemake
```

Or, if you make the WORKDIR suggestion I added to the Dockerfile:

```
docker run --mount type=bind,target=/home/rstudio,source=<LOCAL_PATH_TO_REFINEBIO-EXAMPLES_REPO> -e  refinebio-examples snakemake
```

jashapiro replied:
Even easier, if you are in the refinebio-examples directory: 
```
docker run --mount type=bind,target=/home/rstudio,source=$PWD refinebio-examples snakemake
```

cansavvy replied:
Awesome! I knew there was a way to simplify this but wasn't sure how.

jashapiro replied:
I realized this may actually need to be 
```
docker run --mount type=bind,target=/home/rstudio,source=$PWD refinebio-examples snakemake --cores 1
```
If we use current `snakemake` & other python packages. (which we should)

jashapiro commented on Dockerfile:
```diff
@@ -0,0 +1,75 @@
+# Rtsne and umap are required for dimension reduction analyses
+RUN install2.r --error --deps TRUE \
+    ggfortify \
+    ggsignif \
```
```suggestion
    ggsignif \
    patchwork \
```

cansavvy commented on Dockerfile:
```diff
@@ -0,0 +1,79 @@
+FROM rocker/tidyverse:4.0.0
```
:wow: Okay.

jashapiro commented on CONTRIBUTING.md:
````diff
@@ -1,6 +1,30 @@
+docker pull ccdl/refinebio-examples
+```
+
+Navigate to your local `refinebio-examples` repository.
````
This should happen before they try to `docker build` if they are going that route. Also, if you are there, then you can use:
```
docker run --mount type=bind,target=/home/rstudio,source=$PWD -e PASSWORD=<PASSWORD> -p 8787:8787 refinebio-examples
```
to provide less that needs to be replaced.

jashapiro commented on Dockerfile:
```diff
@@ -0,0 +1,77 @@
+##########################
+
+# Install python3 data science tools
+RUN pip3 install \
+    "cycler==0.10.0" "kiwisolver==1.1.0" "pyparsing==2.4.5" "python-dateutil==2.8.1" "pytz==2019.3" \
+    "numpy==1.17.3" \
+    "pandas==0.25.3" \
+    "snakemake==5.8.1"
```
Updating to current versions...
```suggestion
RUN pip3 install \
    "cycler==0.10.0" "kiwisolver==1.2.0" "pyparsing==2.4.7" "python-dateutil==2.8.1" "pytz==2020.1" \
    "numpy==1.19.1" \
    "pandas==1.1.0" \
    "snakemake==5.20.1"
```

jashapiro commented on Dockerfile:
```diff
@@ -0,0 +1,77 @@
+# Install pip3 and instalation tools
+RUN apt-get -y --no-install-recommends install \
+    python3-pip  python3-dev
+RUN pip3 install "setuptools==46.3.0" "six==1.14.0" "wheel==0.34.2"
```
More version updates
```suggestion
RUN pip3 install "setuptools==49.3.0" "six==1.15.0" "wheel==0.34.2"
```

jashapiro commented on CONTRIBUTING.md:
````diff
@@ -8,22 +8,20 @@ Development should take place in the Docker container.
+Then run this command to start up a container:
 ```
-docker run --mount type=bind,target=/home/rstudio,source=<LOCAL_PATH_TO_REFINEBIO-EXAMPLES_REPO> -e PASSWORD=<PASSWORD> -p 8787:8787 refinebio-examples
+docker run --mount type=bind,target=/home/rstudio,source=$PWD -e PASSWORD=eevee -p 8787:8787 ccdl/refinebio-examples
````
Do you want to note people can change the password?

cansavvy replied:
Shoot! I accidentally copied and pasted that in there.

jashapiro replied:
now everybody knows your secrets 🧐

jashapiro commented on CONTRIBUTING.md:
````diff
@@ -8,22 +8,20 @@ Development should take place in the Docker container.
+docker run --mount type=bind,target=/home/rstudio,source=$PWD -e PASSWORD=eevee -p 8787:8787 ccdl/refinebio-examples
 ```
-Now navigate to http://localhost:8787/ to start developing.
+Now you can navigate to http://localhost:8787/ to start developing.
````
```suggestion
Now you can navigate to http://localhost:8787/ to start developing.
Login to the RStudio server with the username `rstudio` and the password you set above.
```

cansavvy commented on CONTRIBUTING.md:
````diff
@@ -1,6 +1,30 @@
+```
+docker pull ccdl/refinebio-examples
+```
+Replace `<PASSWORD` with your own password in the command below.
````
```suggestion
Replace `<PASSWORD>` with your own password in the command below.
```

cansavvy commented on CONTRIBUTING.md:
````diff
@@ -23,19 +47,21 @@ rule target:
-```
-conda activate snakemake
-```
+Make sure you are running this from a `refinebio-examples` Docker container.
````
```suggestion
Make sure you are running this from a `ccdl/refinebio-examples` Docker container.
```

cansavvy commented:
I believe I've addressed all your comments, @jashapiro . And I've re-built the Docker file and re-ran snakemake succesfully on that new Docker container. 
Is there anything else that needs fixing before we merge? 

It looks like I don't have permission to push this docker image to Docker hub on the CCDL. So that will have to wait I think.

jashapiro commented on CONTRIBUTING.md:
```diff
@@ -96,6 +122,11 @@ Plus its just another thing to have to keep track of.
   - Use "refine.bio", NOT "refinebio"
   - Use `.Rmd`,  NOT "Rmd" or ".Rmd"
   - Use "tidyverse", not "Tidyverse"
+  - Use "TSV" vs tsv or `tsv` or .tsv
+  - Use "PNG" vs png or `png` or .png (and etc.)
```
I heard you like consistent style, so I made a consistent style for your consistent style guide
```suggestion
  - Use "tidyverse", NOT "Tidyverse"
  - Use "TSV",  NOT tsv or `tsv` or .tsv
  - Use "PNG", NOT png or `png` or .png (and etc.)
```

jashapiro commented on 01-getting-started/getting-started.html:
```diff
@@ -1533,11 +1533,11 @@ <h4 class="author">CCDL for ALSF</h4>
 <!-- END doctoc generated TOC please keep comment here to allow auto update -->
 <div id="about-refine.bio" class="section level2">
-<h2>About refine.bio</h2>
+<h2><span class="header-section-number">0.1</span> About refine.bio</h2>
```
I'm just noticing that the numbering is using `0.1` etc. which is kind of ugly. Not needed to be fixed for this PR, but should probably be looked into.

cansavvy replied:
Oh. I kinda liked it like that. But yes. We can change it if we don't like.

jashapiro commented on CONTRIBUTING.md:
````diff
@@ -23,19 +47,21 @@ rule target:
 ```
 snakemake --cores 1
 ```
 If ran successfully, it will spit out a log and all the `.html` output files will have nicely rendered citations.
````
```suggestion
If `snakemake` ran successfully, it will spit out a log and all the `.html` output files will have nicely rendered citations.
```
